### ISSUE TITLE
v137 roadmap

### ISSUE THREAD

**Issue by ije** [esm-dev/esm.sh#932]
## Compile esm.sh to a local CLI that builds esm on your computer

This is useful for people don't like load JS files from a CDN, but want the esm version of a CJS module on NPM.
This idea was discussed on bsky:
- https://jvns.ca/blog/2024/11/18/how-to-import-a-javascript-library/
- https://simonwillison.net/2024/Nov/23/without-a-build-system/
- https://bsky.app/profile/jex.me/post/3lbosjoaawk2w

**Usage:**
```bash
npx esm.sh build --dir=vendor react react-dom
```
then you can import modules like:
```js
import { useState } from "./vendor/react.mjs"
import { createRoot } from "./vendor/react-dom/client.mjs"
```
with importmap:
```js
import { useState } from "react" // "./vendor/react.mjs"
import { createRoot } from "react-dom/client" // "./vendor/react-dom/client.mjs"
```
**Status:**
 - 🚧 on the plan

**Comment by mashaal** [comment]
Nice, I have been doing this manually by digging through the CDN and curling the required files. It might be nice to output the correct import map from this command... either as a file, or in the cli output for easy copying?

**Comment by ije** [comment]
@mashaal i agree, i think the cli should update the root `index.html` by default if it exists, and it also allow you to specify it via flags say `--importmap=**/*.html`

**Comment by b3nten** [comment]
This is a great idea! Would also be nice to install it with Go, and output/modify JSON import map files.

**Comment by dburles** [comment]
This looks amazing. @ije It's similar to what you mentioned here https://github.com/esm-dev/esm.sh/discussions/595#discussioncomment-9977163 a while back. Will the build command support the same package parameters as the CDN? and how will dependency resolution be managed?

**Comment by ije** [comment]
@dburles i think there will be a _package.json_ like file, the CLI should resolve dependencies automatically.

**Comment by guybedford** [comment]
Great to see this as a fundamental goal. Ever since JSPM went CDN-only I've also been wanting to bring back a `--local` flag to the JSPM CLI, ideally with a provider hook to generate a package fully locally and store it in a standard location (formerly this was `jspm_pacakges/registry/package@version`, which could even be brought back).

This is a fundamental security property for JSPM in being able to provide auditable providers as a function of the original npm registry package.

JSPM does also already have the concept of layers on registries, which could possibly incorporate build version / custom build options - `jspm_packages/registry/layer/package@version`, where `layer` might be say `v126-[build-opts-hash]`.

Would be happy to prototype some work here if there is interest in collaboration further too.

**Comment by louwers** [comment]
Any plans to bring the Build API back?
